[Merged by Bors] - feat: `LinearOrder` instance for `Empty`

YanYablonovskiy opened the pull request:
Adds a LinearOrder instance on Empty, is required for defining order types, and complements the PUnit LinearOrder instance below it.


Edit: Refactors LinearOrder on PUnit , and defined an instance on Empty and PEmpty via an instance on Subsingleton.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

When merging, all the commits will be squashed into a single commit
listing all co-authors.

Co-authors in the squash commit are gathered from two sources:

First, all authors of commits to this PR branch are included. Thus,
one way to add co-authors is to include at least one commit authored by
each co-author among the commits in the pull request. If necessary, you
may create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

Second, co-authors can also be listed in lines at the very bottom of
the commit message (that is, directly before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines
at the bottom of the commit message (before the `---`, and also before
any "Co-authored-by" lines) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->
[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

YanYablonovskiy committed:
Update Basic.lean

YaelDillies reviewed (approved):
Thanks!

maintainer delegate

YanYablonovskiy committed:
Apply suggestions from code review

Co-authored-by: Yaël Yanis Dillies <yael.dillies@gmail.com>

YanYablonovskiy committed:
Merge branch 'leanprover-community:master' into Empty.instLinearOrder

YanYablonovskiy committed:
Added `Empty.linearOrder α ` for any `IsEmpty α`

Added `Empty.linearOrder` with signature `(α : Type*) -> [inst : IsEmpty α] -> LinearOrder α` , in order to register instances of `LinearOrder` on `Empty` and `PEmpty`.

YaelDillies commented on Mathlib/Order/Basic.lean:
```diff
@@ -1156,6 +1156,21 @@ lemma eq_or_eq_or_eq_of_forall_not_lt_lt [LinearOrder α]
+  le_total x _ := (inst.false x).elim
+  toDecidableLE x _ := Decidable.isFalse (inst.false x).elim
+
+attribute [instance] linearOrder Empty
```
This makes `linearOrder` and `Empty` into instances. Definitely not what you want!

YanYablonovskiy committed:
Update Mathlib/Order/Basic.lean

Co-authored-by: Yaël Yanis Dillies <yael.dillies@gmail.com>

YanYablonovskiy committed:
Fix gaff

YaelDillies commented on Mathlib/Order/Basic.lean:
```diff
@@ -1156,6 +1156,20 @@ lemma eq_or_eq_or_eq_of_forall_not_lt_lt [LinearOrder α]
 
+namespace Empty
+
+abbrev linearOrder (α : Type*) [inst : IsEmpty α] : LinearOrder α where
```
```suggestion
abbrev LinearOrder.ofIsEmpty (α : Type*) [inst : IsEmpty α] : LinearOrder α where
```
out of the `Empty` namespace, or even generalise to 
```suggestion
abbrev LinearOrder.ofSubsingleton (α : Type*) [Subsingleton α] : LinearOrder α where
```

YanYablonovskiy committed:
Apply suggestions from code review

Co-authored-by: Yaël Yanis Dillies <yael.dillies@gmail.com>

YaelDillies commented on Mathlib/Order/Basic.lean:
```diff
@@ -1156,6 +1156,20 @@ lemma eq_or_eq_or_eq_of_forall_not_lt_lt [LinearOrder α]
+  le_total x _ := (inst.false x).elim
+  toDecidableLE x _ := Decidable.isFalse (inst.false x).elim
+
+instance : LinearOrder Empty := .ofSubsingleton
```
Can you also add `LinearOrder PEmpty` and refactor `LinearOrder PUnit`?

YanYablonovskiy committed:
Refactor `LinearOrder` instances on `Empty`/`Unit` via `Subsingleton`

YanYablonovskiy committed:
Add docstring + fix of trivial for tactic mode

YanYablonovskiy committed:
Reverse unneeded changes due to changed synth order + fixes

YanYablonovskiy committed:
Add `PEmpty` instance

YanYablonovskiy committed:
Getting rid of unneeded tactics

YanYablonovskiy committed:
Reverse changes due to / and  further refactoring of `LinearOrder.ofSubsingleton`

YaelDillies reviewed (approved):
Thanks! 🚀

maintainer merge

YanYablonovskiy committed:
Apply suggestions from code review

Co-authored-by: Yaël Yanis Dillies <yael.dillies@gmail.com>

YanYablonovskiy committed:
Fix a few typos

YanYablonovskiy committed:
Fix spacing

YanYablonovskiy committed:
Remove redundant class assumption

YaelDillies commented on Mathlib/Order/Basic.lean:
```diff
@@ -1156,23 +1156,25 @@ lemma eq_or_eq_or_eq_of_forall_not_lt_lt [LinearOrder α]
+  le_trans x y z _ _ := trivial
+  le_antisymm x y _ _ := Subsingleton.elim x y
+  le_total _ _ := Or.inl trivial
+  lt_iff_le_not_ge _ _ := ⟨fun hf ↦ hf.elim, fun h ↦ h.2 h.1⟩
```
```suggestion
  lt_iff_le_not_ge _ _ := by simp
```
hopefully?

YanYablonovskiy committed:
Apply suggestions from code review

Co-authored-by: Yaël Yanis Dillies <yael.dillies@gmail.com>
